File: 구름 알고리즘 미션/그래프 탐색(1)/택시 기사 구름이/택시 기사 구름이.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const [N, M] = input[0].split(" ").map(Number);
const [X, Y, Z] = input[1].split(" ").map(Number);
const goormMap = input.slice(2, 2 + N).map(line => line.split(" ").map(Number));
const customer = input.slice(2 + N).map(line => line.split(" ").map(Number));

const directions = [[-1, 0],[0, -1],[1, 0],[0, 1]];

const bfs = (start, end) => {
  const [sx, sy] = start;
  const [ex, ey] = end;
  const distance = Array.from({ length: N }, () => Array(N).fill(-1));
  const queue = [[sx, sy]];
  distance[sx][sy] = 0;

  while (queue.length > 0) {
    const [cx, cy] = queue.shift();

    for (const [dx, dy] of directions) {
      const nx = cx + dx;
      const ny = cy + dy;

      if (
        0 <= nx && nx < N &&
        0 <= ny && ny < N &&
        goormMap[nx][ny] !== 1 &&
        distance[nx][ny] === -1
      ) {
        distance[nx][ny] = distance[cx][cy] + 1;
        queue.push([nx, ny]);
      }
    }
  }

  return distance[ex][ey];
}

let [cy, cx] = [customer[0][1] - 1, customer[0][0] - 1];

let result = 0;
for (let i = 0; i < M; i++) {
  let [sy, sx, ey, ex] = customer[i].map(v => v - 1);
  
  let moveDis = 0;
  if (i !== 0) {
    moveDis = bfs([cx, cy], [sx, sy]);
  }
  const workDis = bfs([sx, sy], [ex, ey]);

  let charge = workDis <= 5 ? X : X + (workDis - 5) * Y;

  result += charge - (moveDis + workDis) * Z;

  [cy, cx] = [ey, ex];
}

console.log(result);

File: 구름 알고리즘 미션/그래프 탐색(1)/택시 기사 구름이/택시 기사 구름이.py
import sys
from collections import deque
input = sys.stdin.readline

directions = [[-1, 0], [0, -1], [1, 0], [0, 1]]

def bfs(start, end):
	distance = [[-1] * N for _ in range(N)]
	distance[start[0]][start[1]] = 0
	q = deque([start])
	
	while q:
		cx, cy = q.popleft()
		for dx, dy in directions:
			mx, my = cx + dx, cy + dy
			if 0 <= mx < N and 0 <= my < N and goormMap[mx][my] != 1 and distance[mx][my] == -1:
				distance[mx][my] = distance[cx][cy] + 1
				q.append([mx, my])
	return distance[end[0]][end[1]]
	

N, M = map(int, input().split())
X, Y, Z = map(int, input().split())
goormMap = [list(map(int, input().split())) for _ in range(N)]
customer = [list(map(int, input().split())) for _ in range(M)]

result = 0
cy, cx = customer[0][0] - 1, customer[0][1] - 1

for i in range(M):
	sy, sx, ey, ex = [x-1 for x in customer[i]]
	
	if i == 0:
		moveDis = 0
	else:
		moveDis = bfs([cx, cy], [sx, sy])
	workDis = bfs([sx, sy], [ex, ey])

	if workDis <= 5:
		charge = X
	else:
		charge = X + (workDis - 5) * Y
	
	result += (charge - (moveDis + workDis) * Z)
	
	cx, cy = ex, ey
print(result)
	